### ISSUE TITLE
Public API for Building Containers in Macros

### ISSUE THREAD

**Issue by pulsipher** [jump-dev/JuMP.jl#2604]
For extensions, it would be nice to provide a public API for creating JuMP containers in extension defined macros. Currently, this is accomplished in JuMP's macros via the use of `Containers._build_ref_sets` and `Containers.container_code` which is only partially public (enabling flexibility in JuMP for future macro refactoring). However, this proposed addition would be nice as part of that planned refactoring effort to better enable extension macros that use JuMP containers, but do not wish to build upon `@variable`.  This can be the case for creating extension objects where one or more of the following hold:
- the objects are not decision variables (i.e., using something `@variable(m, p in Parameter)` is not intuitive nomenclature)
- the objects do not allow the use of variable information (e.g., `VariableInfo`)
- the objects are not well amendable to the scalar build paradigm (even with the use of sets)
- the objects can be more concisely/intuitively be defined with extension defined symbolic syntax (as enabled by a macro).
This motivation even presents itself within JuMP itself with `@NLparameter`. In InfiniteOpt, this has several use cases with infinite parameter, finite parameter, and parameter function objects (see the later comments on https://github.com/pulsipher/InfiniteOpt.jl/issues/60).

I know this might not be a high priority at the moment, but I just would like to make this feature request for consideration in future refactoring efforts.

**Comment by odow** [comment]
I think this is really asking for the `Containers` module to be a bit more public and extendable. For example, this needs documenting:
https://github.com/jump-dev/JuMP.jl/blob/f392b0e9e886744bc5b0d9b121c82d0bbf27912d/src/Containers/macro.jl#L201-L204

Ideally, you would only interact with the code in `.Containers`. You shouldn't need code in `src/variables.jl` if you're adding new parameter types, for example.

**Comment by pulsipher** [comment]
> I think this is really asking for the `Containers` module to be a bit more public and extendable. For example, this needs documenting:
> https://github.com/jump-dev/JuMP.jl/blob/f392b0e9e886744bc5b0d9b121c82d0bbf27912d/src/Containers/macro.jl#L201-L204
> 

Precisely, I am essentially advocating for a public API in `Containers` that supports user defined extension macros.

> Ideally, you would only interact with the code in `.Containers`. You shouldn't need code in `src/variables.jl` if you're adding new parameter types, for example.

Well put, that's what I have in mind.

**Comment by pulsipher** [comment]
One more point of clarification is that `parse_container` does conveniently wrap `_build_ref_sets` and `container_code`, but it is not readily usable since macros typically need to intercept the index variables and indices to make the object names:
https://github.com/pulsipher/InfiniteOpt.jl/blob/e76ea9758b653e7c14f97166dd2b4dad9d13098e/src/macros.jl#L686-L694
 
Currently, my work around is to implement my own version of `_build_ref_sets` and then later invoke `JuMP.Containers.container_code`.
